text/3470-crabi-v1.md

```diff
@@ -0,0 +1,661 @@
+that "transforms" a type into the `repr(crabi)` equivalent of that type, to avoid
+having to define a separate type. However, this would not handle the problem of
+translating between the types, which in the general case of a "deep"
+translation could require copying arbitrary in-memory data structures. crABI
+does not aim to solve that problem.
+
+The translation of slices and similar uses structs containing pointer/length
+pairs, rather than inlining the pointer and length as separate arguments.
+[As noted above][types], this is typically passed and returned in an efficient
+fashion on major targets. However, in some languages, such as C, this will
+require separately defining a structure and then using that structure. This
+still seems preferable, though, as combining the two into one struct allows for
+uniform handling between arguments, return values, and fields, as well as
+keeping the pointer and length more strongly associated.
+
+@programmerjake made a
+[proposal](https://github.com/rust-lang/rfcs/pull/3470#issuecomment-1674249638)
+([sample usage](https://github.com/rust-lang/rfcs/pull/3470#issuecomment-1674265515))
+to modify the standard impl of `Drop` for `Box` to allow plugging in an
+arbitrary function (via a `BoxDrop` trait), to drop the `Box` as a whole. This
+would be generally useful (e.g. for object pooling), and would then permit
+crABI to define a `box_drop` function that calls an FFI function to free the
+object. If we accepted that proposal, it would make sense to use it to
+represent crABI boxes.
+
+# Prior art
+[prior-art]: #prior-art
+
+Some potential sources of inspiration:
+
+- WebAssembly Interface Types
+- Swift's stable ABI
+- The `abi_stable` crate (which aims for Rust-to-Rust stability, not
+  cross-language interoperation, but it still serves as a useful reference)
+- `stabby`
+- UniFFI
+- Diplomat
+- C++'s various ABIs (and the history of its ABI changes). crABI does not,
+  however, aim for compatibility with or supersetting of any particular C++
+  ABI.
+- Many, many interface description languages (IDLs).
+- The [x86-64 psABI](https://gitlab.com/x86-psABIs/x86-64-ABI). While we're not
+  specifying the lowering all the way to specific architectures, we can still
+  learn from how it handles various types.
+- The [aarch64 ABI](https://github.com/ARM-software/abi-aa/).
+- The ABIs of other targets.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Once we have support for extern types (such as via
+  https://github.com/rust-lang/rfcs/pull/3396), pointers and references to
+  extern types should work in crABI, just like any other opaque pointer.
+
+- Should the `improper_crabi_types` lint be a hard error, instead?
+
+- Is there a better way we can handle tuple types? Having to use a distinct
+  syntax like `cr#()` is onerous; one of the primary values of tuples is
+  brevity. In the future, if we have variadic generics, we could potentially
```

Comment:
Or, converserly,  variadic generics could build on the crabi tuples. There is definitely some overlap here, since having a tuple with well defined structure, with fields in the declaration order is also useful for using recursion to process the first element of a tuple, and pass the rest of the tuple to next iteration.

Reply:
As you note, there is some overlap, in that both crabi tuples and 'a tuple where you can borrow the tail by reference' (which could potentially used for variadic generics) need the fields to be stored in order.

However, in other ways they have opposite goals.  'A tuple where you can borrow the tail by reference' would likely need to store, e.g., `cr#(u8, u8, u32)` like `cr#(u8, (u8, u32))`.  First, this would be less efficient due to additional padding introduced.  While crABI does already sacrifice layout efficiency to some extent, by giving up reordering and some niche optimizations, that doesn't mean it doesn't care about performance, and this would be an unnecessary extra loss.  Second, given crABI's goal of being easy to bind to other languages, it's beneficial for `cr#(u8, u8, u32)` to translate directly to the obvious C struct equivalent (`struct { uint8_t a; uint8_t b; uint32_t c; }`), rather than to some other structure.